### ISSUE TITLE
Offer flexibility to limit load caused by video transcoding

### ISSUE THREAD

**Issue by marians** [mastodon/mastodon#20064]
### Pitch

We should have better control over

- How many videos are transcoded in parallel
- Timeout for video transcoding jobs
- Exact ffmpeg command executed (point to custom binary or wrapper script)

to limit the danger of DoS effects of mutliple parallel video uploads.

### Motivation

Recently we experienced high load on our instance (running on a single 2 core machine) which caused inability to upload media to new posts in the web UI and also slowed down things notably in general.

We saw that there were 5 to 6 ffmpeg processes running in parallel. The entire load phase lasted for more than one hour in the worst case.

Here is a visualization of the CPU load during that time.

<img width="684" alt="grafik" src="https://user-images.githubusercontent.com/273727/200408222-75d41e4b-f943-43fe-89e9-992d3d29537c.png">

Unfortunately I was unable to find any means to control these jobs.

I checked the sidekiq dashboard but didn't find anything related. This made me wonder: Are video transcoding jobs even managed through sidekiq? The process tree shown by `ps faux` indicates to me that the `ffmpeg` commands are not executed through sidekiq workers. Which makes we wonder why. Ideally these would be sidekiq jobs in a separate queue which we could offload to a separate machine.

<img width="1205" alt="grafik" src="https://user-images.githubusercontent.com/273727/200410132-c70d1df2-024d-434d-b007-682710edea37.png">

It occurs to me that a simple concurrency limit configuration for this type of job would make sense, so that the likelihood to fully saturate the CPU can be reduced.

Also, limiting the execution time (and then fail the job without retry) would mean that videos that passed the uplaod size limit (which I found to be 40 MB), but still are too complex to transcode, would not block the instance.

Lastly, it would be benefitial if we could customize how `ffmpeg` is called. For reasons unknown to me there is code to read an `FFMPEG_BINARY` environment variable here:

https://github.com/mastodon/mastodon/blob/v4.0.0rc2/config/initializers/ffmpeg.rb

From my tests, and from looking at how ffmpeg gets executed, this variable has no effect.

Being able to point this variable to a wrapper script, to apply tools like `cpulimit` and `nice`, seems like a good way to give a lot more power to admins to limit the blast radius.

Mastodon is becoming more important. Let's also reduce its attack surface and resource consumption where reasonably possible.

**Comment by marians** [comment]
Related: https://github.com/mastodon/mastodon/issues/8648

**Comment by vielmetti** [comment]
Related: #16496

**Comment by marians** [comment]
@vielmetti Thanks for the link! Indeed I consider this a

Duplicate of #16496
